Fix #12802 - allow to override preference over CartInterface and return correct object from QuoteRepository

Bartlomiejsz opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
This is fix for #12802 

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento2#12802: QuoteRepository get methods won't return CartInterface but Quote model

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
Steps are described in #12802 

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

orlangur commented on app/code/Magento/Quote/Model/QuoteRepository.php:
```diff
@@ -203,13 +214,13 @@ public function delete(\Magento\Quote\Api\Data\CartInterface $quote)
-        if ($sharedStoreIds) {
+        /** @var CartInterface $quote */
+        $quote = $this->cartFactory->create();
+        if ($sharedStoreIds && $quote instanceof Quote) {
```
Maybe just check `method_exists('setSharedStoreIds')` ?

Bartlomiejsz replied:
@orlangur actually this method does not exist I believe, it is declared only in `Quote` class docblock. Maybe we should change this to `$quote instanceof \Magento\Framework\DataObject || method_exists($quote, 'setSharedStoreIds')`? Because in both situations method should probably be called.

Bartlomiejsz replied:
@orlangur please check current solution 😉

orlangur replied:
Please grep for `share.*store` to find whether we get it somewhere.

It works and is placed in DocBlock due to `DataObject` providing magic method `__call`. I would use it for a legacy `Quote` only or would check with `method_exists` so it won't break any implementation just covering `CartInterface` but not extending `AbstractModel`.

Bartlomiejsz replied:
I haven't found any usage that could be broken using grep. Please check now, I changed DataModel to Quote @orlangur

orlangur replied:
I meant remove this condition if no getter called but this is not true: https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Quote/Model/ResourceModel/Quote.php#L63

Bartlomiejsz replied:
@orlangur what should I do with this condition then?

orlangur reviewed (changes requested):
Really nice @Bartlomiejsz!

According to my understanding such change is backward compatible, asking @sivaschenko to double check.

Please mark `quoteFactory` property as deprecated.

After changes are applied and all builds are green, please squash them into a single commit so that we have perfectly clean history 😉

sivaschenko commented on app/code/Magento/Quote/Model/QuoteRepository.php:
```diff
@@ -77,39 +81,47 @@ class QuoteRepository implements \Magento\Quote\Api\CartRepositoryInterface
         StoreManagerInterface $storeManager,
-        \Magento\Quote\Model\ResourceModel\Quote\Collection $quoteCollection,
-        \Magento\Quote\Api\Data\CartSearchResultsInterfaceFactory $searchResultsDataFactory,
-        JoinProcessorInterface $extensionAttributesJoinProcessor,
```
I don't think it's a good idea to change names of the parameters, as they could be referenced i.e. in arguments DI configuration.

Bartlomiejsz replied:
@sivaschenko previously Codacy result was failed due to length of those, should I ignore this then and change them back to length not accepted by Codacy?

orlangur replied:
@Bartlomiejsz yep, feel free to ignore any of those, as we removed such check for PHPMD intentionally. I heard Codacy is going to be eliminated soon.

Bartlomiejsz replied:
@sivaschenko @orlangur done

sivaschenko reviewed (changes requested):
Hi @Bartlomiejsz thanks for the pull request, please see my review comment

orlangur commented on app/code/Magento/Quote/Model/QuoteRepository.php:
```diff
@@ -203,13 +215,15 @@ public function delete(\Magento\Quote\Api\Data\CartInterface $quote)
+        /** @var CartInterface $quote */
+        $quote = $this->cartFactory->create();
+        if ($sharedStoreIds &&
+            ($quote instanceof Quote || method_exists($quote, 'setSharedStoreIds'))
```
Remove first part of condition. `method_exists` is sufficient.

Bartlomiejsz replied:
@orlangur done

orlangur reviewed (changes requested):
After changes are applied and all builds are green, please squash them into a single commit so that we have perfectly clean history 😉

Bartlomiejsz committed:
Fix #12802 - allow to override preference over \Magento\Quote\Api\Data\CartInterface and return correct object from QuoteRepository (+4 squashed commits)
Squashed commits:
[55b9f3ec52b] Fix #12802 - fix phpmd, mark quoteFactory as deprecated
[56ca9a42468] Fix #12802 - change condition in quoteRepository
[734212812a4] Fix #12802 - revert change of constructor parameters names
[ba8ad543e0f] Fix #12802 - remove instanceof condition

Bartlomiejsz commented:
@orlangur commits are squashed now

magento-engcom-team commented:
Hi @orlangur, thank you for the review.
[ENGCOM-4679](https://jira.corp.magento.com/browse/ENGCOM-4679) has been created to process this Pull Request

nmalevanec committed:
Merge remote-tracking branch 'mainline/2.3-develop' into feature/fix_12802_quote_not_overridable

nmalevanec committed:
Fix static tests.
